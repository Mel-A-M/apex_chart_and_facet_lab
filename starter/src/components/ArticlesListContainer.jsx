/**
 * Copyright (c) 2020 Oracle and/or its affiliates. All rights reserved.
 * Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl.
 */

import React from 'react';
import PropTypes from 'prop-types';

import getDeliveryClient from '../scripts/server-config-utils';
import { fetchArticles } from '../scripts/services';

import ArticlesList from './ArticlesList';

/**
 * Component responsible for getting the list of articles for a topic.
 *
 * Note: This actual data is rendered in the "ArticlesList",
 * this design is to keep the model and view separate.
 *
 * @param {string} topicId The ID of the topic whose articles are to be obtained
 * @param {string} topicName The Name of the topic whose articles are to be obtained
 */
export default class ArticlesListContainer extends React.Component {
  constructor(props) {
    super(props);

    // keep track of whether this component is mounted or not
    this.mounted = false;

    // the component's state
    this.state = {
      loading: true,
      error: false,
      articles: {},
    };
  }

  /**
   * Load the data for the specific topic
   */
  componentDidMount() {
    this.mounted = true;

    // set the browser tab title
    document.title = 'Articles';

    // get the client to connect to CEC
    const deliveryClient = getDeliveryClient();

    const { topicId } = this.props;

    // fetch the articles for the topic
    fetchArticles(deliveryClient, topicId)
      .then((articles) => {
        if (this.mounted) {
          this.setState({ articles });
        }
      })
      .catch(() => { this.setState({ error: true }); })
      .then(() => this.setState({ loading: false }));
  }

  /*
   * Called when the component unmounts.
   */
  componentWillUnmount() {
    this.mounted = false;
  }

  /*
   * Render an error if the topic was not found.
   * If the found was obtained ok render the "ArticlesList" component
   */
  render() {
    const { topicId, topicName } = this.props;
    const { error, loading, articles } = this.state;

    return (
      <div data-testid="ArticlesListContainer">
        {/* Render error */}
        {error && (
          <p>
            <h2>Topic Not Found</h2>
            <i>Unable to list articles.</i>
          </p>
        )}

        {/* Render loading */}
        {loading && !error && (
          <div className="progress-spinner" />
        )}

        {/* Render data */}
        {!error && !loading && (
          <ArticlesList
            articles={articles}
            topicId={topicId}
            topicName={topicName}
          />
        )}
      </div>
    );
  }
}

/*
 * Define the type of data used in this component.
 */
ArticlesListContainer.propTypes = {
  topicId: PropTypes.string.isRequired,
  topicName: PropTypes.string.isRequired,
};
