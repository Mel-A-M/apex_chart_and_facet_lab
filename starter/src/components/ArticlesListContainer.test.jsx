/* eslint-disable no-undef */
import React from 'react';
import { waitForElement, render } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import ArticlesListContainer from './ArticlesListContainer';

import MockServices from '../../test/MockServices';

/** **********************************************************
 * Perform validation of the ArticlesListContainer component
 *
 * Data is provided from mocked calls to the service wrapper
 * API
 */
describe('Validation of the ArticlesListContainer component', () => {
  beforeEach(() => {
    // Use the predefined mock responses for all service wrapper methods
    MockServices.overrideServiceWrapper();
  });

  afterEach(() => {
    // Reset the mock resonses so as not to interfere with other tests
    MockServices.resetServiceWrapper();
  });

  /** ****************************************
   * Perform snapshot test
   ***************************************** */
  test('Render the articles list for a specific container', async () => {
    /** *************
     * Arrange
     ************** */
    expect.assertions(1);

    /** *************
     * Act
     ************** */
    const { container, getByTestId } = render(
      <Router path="/">
        <ArticlesListContainer
          topicId="CORE3E84C4A3FC084704B0EDE88515BB4506"
          topicName="How%20To"
        />
      </Router>,
    );

    /** *************
     * Assert
     ************** */
    await waitForElement(() => getByTestId('ArticlesList'));

    expect(container).toMatchSnapshot();
  });
});
