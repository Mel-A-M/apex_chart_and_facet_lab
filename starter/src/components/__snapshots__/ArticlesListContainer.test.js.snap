// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validation of the ArticlesListContainer component Render the articles list for a specific container 1`] = `
<div>
  <div
    data-testid="ArticlesListContainer"
  >
    <div
      data-testid="ArticlesList"
    >
      <div
        id="breadcrumb"
      >
        <ul>
          <li>
            <a
              href="/"
            >
              Home
            </a>
          </li>
          <li>
            How%20To
          </li>
        </ul>
      </div>
      <div
        id="articles"
      >
        <div>
          <div
            class="article"
          >
            <a
              href="/article/COREC290E8ABC4354431863218B6C974122B?topicName=How%20To&topicId=CORE3E84C4A3FC084704B0EDE88515BB4506"
              style="text-decoration: none;"
            >
              <div
                class="title-date"
              >
                <h4
                  class="title"
                >
                  Create Beautiful Latte Art!
                </h4>
                <div
                  class="date"
                >
                  Posted on May 8, 2019
                </div>
              </div>
              <img
                src="http://www.somewhere.com/CONT3C779E28717540048AB2C3EF7BEFBFC7.png"
              />
              <div
                class="description"
              >
                With a little practice, you can too can create designs with steamed milk.
              </div>
            </a>
          </div>
        </div>
        <div>
          <div
            class="article"
          >
            <a
              href="/article/COREED6BA23517A148E3A956DCB5A5D39256?topicName=How%20To&topicId=CORE3E84C4A3FC084704B0EDE88515BB4506"
              style="text-decoration: none;"
            >
              <div
                class="title-date"
              >
                <h4
                  class="title"
                >
                  Pouring Just The Right Cup
                </h4>
                <div
                  class="date"
                >
                  Posted on May 8, 2019
                </div>
              </div>
              <img
                src="http://www.somewhere.com/CONT182BF086A70341F3AC90F81F2621903A.png"
              />
              <div
                class="description"
              />
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
